### PR TITLE
Update genesis to start with Is Trigger

### PR DESCRIPTION
**Pull request by shayzluf**
Implement Is genesis trigger method for genesis state creation

### CODE REVIEW

File: beacon-chain/powchain/log_processing.go [hunk 1]
```diff
@@ -211,8 +219,15 @@ func (w *Web3Service) processPastLogs() error {
 		return err
 	}
 
-	for _, log := range logs {
-		w.ProcessLog(log)
+	for _, l := range logs {
+		if err := w.ProcessLog(l); err != nil {
+			log.WithError(err).Error("Could not process log")
+		}
+		if w.chainStarted {
```
[the remark this answers is not in the document]
Comment (shayzluf) [reply beacon-chain/powchain/log_processing.go]: moved

File: beacon-chain/powchain/log_processing.go [hunk 2]
```diff
@@ -248,11 +263,17 @@ func (w *Web3Service) requestBatchedLogs() error {
 	// Only process log slices which are larger than zero.
 	if len(logs) > 0 {
 		log.Debug("Processing Batched Logs")
-		for _, log := range logs {
-			w.ProcessLog(log)
+		for _, l := range logs {
+			if err := w.ProcessLog(l); err != nil {
+				log.WithError(err).Error("Could not process log")
+			}
+			if w.chainStarted {
```
[the remark this answers is not in the document]
Comment (shayzluf) [reply beacon-chain/powchain/log_processing.go]: removed

File: beacon-chain/blockchain/service_test.go [hunk 3]
```diff
@@ -218,6 +218,15 @@ func SetSlotInState(service *ChainService, slot uint64) error {
 	return service.beaconDB.SaveState(context.Background(), bState)
 }
 
+func setLatestStateRootsInState(service *ChainService) error {
+	bState, err := service.beaconDB.HeadState(context.Background())
```
[the remark this answers is not in the document]
Comment (shayzluf) [reply beacon-chain/blockchain/service_test.go]: moved

File: beacon-chain/blockchain/service.go [hunk 4]
```diff
@@ -7,6 +7,8 @@ import (
 	"bytes"
 	"context"
 	"fmt"
+	"github.com/prysmaticlabs/go-ssz"
```
Comment (terencechain) [inline beacon-chain/blockchain/service.go]: run go imports

File: beacon-chain/core/state/state.go [hunk 5]
```diff
@@ -112,21 +91,93 @@ func GenesisBeaconState(deposits []*pb.Deposit, genesisTime uint64, eth1Data *pb
...
+//            validator.activation_epoch = GENESIS_EPOCH
+//
+//    # Populate latest_active_index_roots
+//    genesis_active_index_root = hash_tree_root(get_active_validator_indices(state, GENESIS_EPOCH))
+//    for index in range(LATEST_ACTIVE_INDEX_ROOTS_LENGTH):
+//        state.latest_active_index_roots[index] = genesis_active_index_root
+//
+//    return state
+
+// def initialize_beacon_state_from_eth1(eth1_block_hash: Hash,
```
Comment (terencechain) [inline beacon-chain/core/state/state.go]: why's there two functions embedded in the comments?

File: beacon-chain/powchain/log_processing.go [hunk 6]
```diff
@@ -153,26 +144,48 @@ func (w *Web3Service) ProcessDepositLog(depositLog gethTypes.Log) {
...
-// the ETH1.0 chain by trying to determine when to start the beacon chain.
-func (w *Web3Service) ProcessChainStartLog(depositLog gethTypes.Log) {
-	chainStartCount.Inc()
-	chainStartDepositRoot, _, timestampData, err := contracts.UnpackChainStartLogData(depositLog.Data)
-	if err != nil {
-		log.Errorf("Unable to unpack ChainStart log data %v", err)
-		return
+// isGenesisTrigger gets called whenever there's a deposit event,
+// it checks whether there's enough effective balance to trigger genesis.
+// Spec pseudocode definition:
```
Comment (terencechain) [inline beacon-chain/powchain/log_processing.go]: you need a space before `// Spec pseudocode definition:`

File: beacon-chain/core/state/state.go [hunk 7]
```diff
@@ -5,45 +5,17 @@ package state
...
-//            validator.activation_epoch = GENESIS_EPOCH
-//
-//    genesis_active_index_root = hash_tree_root(get_active_validator_indices(state, GENESIS_EPOCH))
-//    for index in range(LATEST_ACTIVE_INDEX_ROOTS_LENGTH):
-//        state.latest_active_index_roots[index] = genesis_active_index_root
-//
-//    return state
-func GenesisBeaconState(deposits []*pb.Deposit, genesisTime uint64, eth1Data *pb.Eth1Data) (*pb.BeaconState, error) {
+// BeaconState gets called to return a beacon state where all the fields are set to genesis values.
+func BeaconState(blkHeader *pb.BeaconBlockHeader, genesisTime uint64, eth1Data *pb.Eth1Data) *pb.BeaconState {
```
Comment (terencechain) [inline beacon-chain/core/state/state.go]: what's the difference between `BeaconState` and `GenesisBeaconState`? Why aren't they together? We only need to call `BeaconState` once

File: beacon-chain/core/state/state.go (hunk 5 above)
Comment (shayzluf) [reply beacon-chain/core/state/state.go]: removed

File: beacon-chain/core/state/state.go (hunk 7 above)
Comment (shayzluf) [reply beacon-chain/core/state/state.go]: it becomes a vary long function if we don't separate the initialization from the computation

File: beacon-chain/blockchain/service.go [hunk 8]
```diff
@@ -139,12 +140,13 @@ func (c *ChainService) initializeBeaconChain(genesisTime time.Time, deposits []*
 	if err := c.beaconDB.InitializeState(c.ctx, unixTime, deposits, eth1data); err != nil {
 		return nil, fmt.Errorf("could not initialize beacon state to disk: %v", err)
 	}
-	beaconState, err := c.beaconDB.HeadState(c.ctx)
+
+	beaconState, err := state.GenesisBeaconState(deposits, unixTime, eth1data)
```
[the remark this answers is not in the document]
Comment (nisdas) [reply beacon-chain/blockchain/service.go]: do you not agree with this ? I still am not clear why we are creating it again here

File: beacon-chain/core/state/state.go (hunk 7 above)
Comment (nisdas) [reply beacon-chain/core/state/state.go]: I am not sure what you mean by this, why cant we do it the same way we have been doing as before ?

File: beacon-chain/powchain/log_processing.go (hunk 6 above)
Comment (terencechain) [reply beacon-chain/powchain/log_processing.go]: sorry I meant a new line

File: beacon-chain/core/state/state.go [hunk 9]
```diff
@@ -5,45 +5,17 @@ package state
...
-//        if validator.effective_balance >= MAX_EFFECTIVE_BALANCE:
-//            validator.activation_eligibility_epoch = GENESIS_EPOCH
-//            validator.activation_epoch = GENESIS_EPOCH
-//
-//    genesis_active_index_root = hash_tree_root(get_active_validator_indices(state, GENESIS_EPOCH))
-//    for index in range(LATEST_ACTIVE_INDEX_ROOTS_LENGTH):
-//        state.latest_active_index_roots[index] = genesis_active_index_root
-//
-//    return state
-func GenesisBeaconState(deposits []*pb.Deposit, genesisTime uint64, eth1Data *pb.Eth1Data) (*pb.BeaconState, error) {
```
[the remark this answers is not in the document]
Comment (shayzluf) [reply beacon-chain/core/state/state.go]: removed it

File: beacon-chain/blockchain/service.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (shayzluf) [reply beacon-chain/blockchain/service.go]: removed it

File: beacon-chain/core/state/state.go (hunk 7 above)
Comment (shayzluf) [reply beacon-chain/core/state/state.go]: removed

**Comment by shayzluf** [comment]
started from scratch in https://github.com/prysmaticlabs/prysm/pull/2905

**Pull request closed without merging**